### ISSUE TITLE
✨ Minimap

### ISSUE THREAD

**Issue by austincondiff** [CodeEditApp/CodeEditSourceEditor#33]
### Description

The Minimap provides a compact visual summary of your code, helping you quickly get a sense of the file’s structure and flow. It appears along the right edge of the editor and lets you instantly jump to different parts of the file by clicking or dragging within the highlighted view area.

Implement the visual display system for the **Minimap** in the source editor. The Minimap should provide a vertically compressed, non-interactive rendering of the entire file's syntax structure. It visually represents the document using **color rectangles**, offering users an overview for orientation and navigation.

### Requirements

- Display a vertical Minimap next to the code editor view
- The Minimap should:
  - Represent the full document using only **color rectangles**, not text
  - Use syntax highlighting colors from the active theme
  - Dynamically reflect updates in the main editor (insert, delete, scroll, etc.)

- Scrolling and navigation:
  - Scrolling the editor updates the Minimap viewport
  - Clicking or dragging in the Minimap scrolls the editor to the corresponding position
  - A translucent viewport rectangle highlights the currently visible section of the document

- Selection handling:
  - If there is **no selection**, highlight the current line
  - If there is a **text selection**, highlight the selected range
  - If there are **multiple selections**, show all selection ranges

- Integration with editor features:
  - Display search result locations (e.g., from Find/Replace)
  - Display error and warning markers (e.g., from diagnostics)
  - Indicate folded regions if any exist

- Behavior and customization:
  - The Minimap should be toggleable in editor settings
 
### Implementation Details

- Rendering should be GPU-accelerated or use offscreen rendering for performance
- Consider lazy rendering or caching for large files to reduce memory and CPU usage
- The Minimap should not allow direct text editing
- Rendering of syntax highlight colors can reuse the existing tokenization model

### Additional Context

This feature provides users with a consistent and performant way to understand the overall structure of their code. It is especially helpful in long files and serves as a complement to navigation and issue reporting systems in the editor.

This lays the foundation for future enhancements such as semantic structure previews, bookmarks, and inline annotations inside the Minimap.

### Screenshots

<img width="899" alt="Image" src="https://github.com/user-attachments/assets/570a5a3e-8d35-4e65-9d64-ef5ba27353d5" />

### Resources

- [VSCode minimap documentation](https://code.visualstudio.com/docs/getstarted/userinterface#_minimap)
- [Xcode's Minimap](https://verbalraj.medium.com/xcodes-minimap-2023d662c2a4)
- [Xcode documentation](https://developer.apple.com/documentation/xcode/configuring-the-xcode-project-window#:~:text=The%20Minimap%20view%20provides%20a%20miniaturized%20version%20of%20your%20source%20files%20content%2C%20which%20you%20use%20to%20navigate%20around%20the%20file.)
- [CodeEditorView minimap source](https://github.com/mchakravarty/CodeEditorView/blob/main/Sources/CodeEditorView/MinimapView.swift)

### Related Issues
- https://github.com/CodeEditApp/CodeEditTextView/pull/84

**Comment by Yellow-GGG** [comment]
Actually I'd prefer the minimap like Xcode rather than VScode. The minimap of Xcode contains more useful features and looks cleaner and much more beautiful. It will definitely take more work to make one like that, but I believe you'll love it too.

**Comment by tjkohli** [comment]
@Yellow-GGG I agree the Xcode minimap is far more useful; however at the same time I think it may be wiser to ship a minimap with minimal functionality and then add complexity in future releases.

**Comment by Eliulm** [comment]
Hey @austincondiff, could you assign this to me? I already got some of the CodeEditorView code working:


![Image](https://user-images.githubusercontent.com/82230675/229353000-fb059930-c18f-4884-8721-e8c49b1001c0.png)

**Comment by Eliulm** [comment]
This is the current state of the minimap. Still a lot of bugs to fix, but I am getting there slowly...

https://user-images.githubusercontent.com/82230675/231145498-c3f3ad40-2a18-4034-9e3c-c883e0843a42.mov

@thecoolwinter I need to set the bounds of the visibleRange in `STTextview+VisibleRange` to the visibleBounds of the MinimapView. How should I rewrite the code, so that I can pass props to `visibleTextRange` variable?

**Comment by austincondiff** [comment]
Looks great! Where are we with this?

**Comment by Eliulm** [comment]
Unfortunately, I have had little time recently, since I am currently working on my degree.

I have decided that I am going re-implement the minimap in textkit2, instead of textkit1, since the resizing was very laggy. I have also encountered multiple issues like #184, which need to be resolved first, so that the minimap can work properly.

**Comment by austincondiff** [comment]
@Eliulm are you willing to finish this? If not can you provide the code so we can pick up where you left off?

**Comment by thecoolwinter** [comment]
I'm going to be taking this on, here's my basic plan (I'll be going a similar direction to CodeEditorView like mentioned above):

The minimap is basically another TextView, except it lays out and renders it's lines using small dots. Because the layout manager is related to the text storage now, we can have the minimap and textview share a storage, but each have their own layout information.

The outline of required changes:
- [x] Create an API to allow for overriding layout and drawing for the layout manager [cetv pr](https://github.com/CodeEditApp/CodeEditTextView/pull/84)
- [ ] Create a Minimap View that tells the layout manager about it's available space, and draws text as small dots
- [ ] Overlay the minimap view on the scrollview in CESE
- [ ] Create a 'viewport' view that tracks the scroll offset of the editor, and matches it on the minimap.
